pkg/image/controller/trigger/trigger_controller.go

```diff
@@ -0,0 +1,399 @@
+	c := &TriggerController{
+		eventRecorder:    eventBroadcaster.NewRecorder(api.EventSource{Component: "image-trigger-controller"}),
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "image-trigger"),
+		imageChangeQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "image-trigger-reactions"),
+		lister:           lister,
+		tagRetriever:     tagRetriever{lister},
+		triggerCache:     NewTriggerCache(),
+
+		resourceFailureDelayFn: defaultResourceFailureDelay,
+	}
+
+	c.syncImageStreamFn = c.syncImageStream
+	c.syncResourceFn = c.syncResource
+	c.enqueueImageStreamFn = c.enqueueImageStream
+
+	isInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addImageStreamNotification,
+		UpdateFunc: c.updateImageStreamNotification,
+	})
+	c.syncs = []cache.InformerSynced{isInformer.HasSynced}
+
+	triggers, syncs, err := setupTriggerSources(c.triggerCache, c.tagRetriever, sources, c.imageChangeQueue)
+	if err != nil {
+		panic(err)
+	}
+	c.triggerSources = triggers
+	c.syncs = append(c.syncs, syncs...)
+
+	return c
+}
+
+func setupTriggerSources(triggerCache cache.ThreadSafeStore, tagRetriever trigger.TagRetriever, sources []TriggerSource, imageChangeQueue workqueue.RateLimitingInterface) (map[string]TriggerSource, []cache.InformerSynced, error) {
+	var syncs []cache.InformerSynced
+	triggerSources := make(map[string]TriggerSource)
+	for _, source := range sources {
+		if source.Store == nil {
+			source.Store = source.Informer.GetStore()
+		}
+		prefix := source.Resource.String() + "/"
+		if _, ok := triggerSources[source.Resource.String()]; ok {
+			return nil, nil, fmt.Errorf("duplicate resource names registered in %#v", sources)
+		}
+		triggerSources[source.Resource.String()] = source
+
+		handler := ProcessEvents(triggerCache, source.TriggerFn(prefix), imageChangeQueue, tagRetriever)
+		source.Informer.AddEventHandler(handler)
+		syncs = append(syncs, source.Informer.HasSynced)
+	}
+	return triggerSources, syncs, nil
+}
+
+// Run begins watching and syncing.
+func (c *TriggerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting trigger controller")
+
+	if !cache.WaitForCacheSync(stopCh, c.syncs...) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.imageStreamWorker, time.Second, stopCh)
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.resourceWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down trigger controller")
+}
+
+func (c *TriggerController) addImageStreamNotification(obj interface{}) {
+	is := obj.(*imageapi.ImageStream)
+	//glog.V(4).Infof("Adding image stream %s", is.Name)
```

Comment:
Maybe, we care about image stream when we can use them in resources so I could see this controller logging these.